Normalize the direction vector passed to Body::intersectsRay()

peci1 opened the pull request:
All code inside intersectsRay silently expects the direction to be unit.

I added an automatic normalization of the passed direction.

peci1 commented:
E.g. `detail::distanceSQR` which is used everywhere except `Box::intersectsRay()` expects unit direction (according to the docs at https://github.com/ros-planning/geometric_shapes/blob/f8ca20902bd3455ef60bdd18420b54141ebefc30/src/bodies.cpp#L77 ).

rhaschke reviewed (commented):
I agree that this assumption should be considered.

> I find this solution better than silently normalizing the vector because that would be more expensive.

However, I think the overhead of throwing inside the function is more costly than performing 3 more divisions. Usually, we try to avoid throwing in the MoveIt code base.

peci1 commented:
You're right that throwing is expensive, but I think it's only a corner case and not something that would happen often...

rhaschke commented:
That's true, we can safe the default exception handling overhead only if declaring the function as `nothrow`. Let's see what other maintainers think of these exceptions. @v4hn, @davetcoleman.

v4hn commented:
@peci1 do you have some benchmark statistics that show the overhead of implicitly normalizing?
I would not expect a tremendous effect and would prefer the implicit normalization in this case.

On the proposed throws, I would argue that this is no case you might ever want to catch anywhere.
I would actually be fine with hard asserts the way OpenCV handles these validations.

davetcoleman commented:
I don't have a strong opinion about these exceptions, seems fine to me

peci1 committed:
Normalize the direction passed to Body::intersectsRay().

peci1 commented:
Well, I convinced myself into going the way of automatic normalization. These functions are expensive anyways. 

Method `MatrixBase<Derived>::normalized()` first computes the squared norm, so just a few multiplications and additions. Then, if the squared norm is not one, it performs the normalization. But there is a comment in the method body saying that the conditional doesn't cause any slowdowns on modern CPUs/compilers. So we can hope that for the case of normalized direction, the additional cost will not be very high.

peci1 commented:
I did the performance comparison in the end. First, I figured out I misread the `.normalized()` function - it only checks, if the norm is zero, not if it is one. so it doesn't help with already normal vectors.

Now the comparison. I was calling 100.000.000 times `Box::intersectsRay()`. I set CPU governor to performance and let the CPU cool down between all tests so that it doesn't get throttled. I repeated each test 3-4 times and averaged off the top of my head.

- Throwing exceptions: avg. 3.9 sec (only unit vetors tried)
- normalizing all incoming vectors: avg. 5.0 sec (both unit and non-unit vectors)
- comparing `squaredNorm()` to 1, normalizing only non-unit vectors: avg. 4.0 sec for unit vectors and 5.0 for non-unit vectors

So in the end I implemented the last variant, where unit vectors get almost no penalization, and non-unit vectors slow down by approx. 25%. That's a lot, so I added a note to the function docs that passing unit vectors will speed the computation up.

v4hn reviewed (approved):
looks good to go in its current state.

rhaschke committed:
factor out normalize() function

rhaschke reviewed (approved):
I factored out the repeated normalization code into a separate function.

rhaschke commented:
@peci1, Having merged some previous PRs, this one has conflicts now. Could you please rebase?

peci1 commented:
We had a discussion with  @v4hn about the use of `Eigen::numext::sqrt` and we concluded that until there are some benchmarks, it is proabably not worth it to spread it in moveit code. But as this is just a small library, I think we can retain it here. It would also be easier to do performance tests with this library than with the whole moveit...

peci1 commented:
I'll rebase.

rhaschke commented:
Can you, maybe advice on the order of merging to avoid these conflicts?

peci1 commented:
Rebased.

peci1 commented:
I think there will always be conflicts. Just go on and ping me whenever a rebase is needed.

rhaschke commented:
@peci1, I commented on two other PRs, which need some more intervention.
